File: PixelSolution/PixelEngine/Transform.cpp
#include "pch.h"
#include "Transform.h"
#include "PixelEngineAPI.h"
#include "Registry.h"
#include "EntityObject.h"
TransformData* Transform_Add(unsigned int id)
{
	auto registry = GetRegistry();
	registry->Add<TransformData>(id);
	registry->Add<WorldData>(id);

	auto data = registry->Get<TransformData>(id);
	data->thisID = id;
	return data;
}

TransformData* Transform_Get(unsigned int id)
{
	auto registry = GetRegistry();
	TransformData* data = registry->Get<TransformData>(id);
	if (data == nullptr)
	{
		PixelLog::Error("[Transform][GetComponent] Not Find Component");
	}
	return data;
}

bool Transform_Has(unsigned int id)
{
	auto registry = GetRegistry();
	TransformData* data = registry->Get<TransformData>(id);
	if (data == nullptr)
	{
		return false;
	}
	else
	{
		return true;
	}
}

void Transform_SetParent(unsigned int id, unsigned int ParentID)
{
	auto thisEntity = FindEntity(id);
	if (thisEntity == nullptr)
	{
		PixelLog::Error("[Transform][SetParent] Not Find This EntityID :" + id);
		return;
	}
	thisEntity->SetParent(ParentID);
}

void Transform_SetChild(unsigned int id, unsigned int ChildID)
{
	auto thisEntity = FindEntity(id);
	if (thisEntity == nullptr)
	{
		PixelLog::Error("[Transform][SetParent] Not Find This EntityID :" + id);
		return;
	}
	thisEntity->SetChild(ChildID);

}

void Transform_SetLocalPosition(unsigned int id, PVector3 position)
{


}

void Transform_SetLocalRotation(unsigned int id, PVector3 rotation)
{

}

void Transform_SetLocalScale(unsigned int id, PVector3 scale)
{

}

File: Survival/Engine/Transform.lua
local ffi = require("ffi")
local dll = ffi.load("PixelEngine")

ffi.cdef[[
	typedef struct {
		PVector3 position;
		unsigned int thisID;
		PVector3 rotation;
		unsigned int bitmask;
		PVector3 scale;
		unsigned int unused2;
	} TransformData;

	TransformData* Transform_Add(unsigned int ID);
	TransformData* Transform_Get(unsigned int ID);
	bool Transform_Has(unsigned int ID);
	void Transform_SetParent(unsigned int ID,unsigned int ParentID);
	void Transform_SetChild(unsigned int ID,unsigned int ChildID);
]]

---@class TransformData
---@field position PVector3
---@field rotation PVector3
---@field scale PVector3
local TransformData = {}

---@param ParentID number
function TransformData:SetParent(ParentID)
	return dll.Transform_SetParent(self.thisID,ParentID)
end

---@param ChildID number
function TransformData:SetChild(ChildID)
	return dll.Transform_SetChild(self.thisID,ChildID)
end

local Transform_mt = {
	__index = TransformData
}

ffi.metatype("TransformData", Transform_mt)

Transform = Transform or {}

---@param ID number 
---@return TransformData 
function Transform.Add(ID) 
	local data = dll.Transform_Add(ID)
	data.thisID = ID
	return data 
end 

---@param ID number 
---@return TransformData 
function Transform.Get(ID) 
	return dll.Transform_Get(ID) 
end 

---@param ID number 
---@return boolean 
function Transform.Has(ID) 
	return dll.Transform_Has(ID) 
end 

return Transform